ignite/metrics/running_average.py

```diff
@@ -139,16 +142,62 @@ def compute(self) -> Union[torch.Tensor, float]:
+                ======================================================= ===========================================
+                :class:`~.metrics.metric.RunningBatchWise`              Running average of the ``src`` metric or
+                                                                        ``engine.state.output`` is computed across
+                                                                        batches. In the former case, on each batch,
+                                                                        ``src`` is reset, updated and computed then
+                                                                        its value is retrieved.
+                :class:`~.metrics.metric.SingleEpochRunningBatchWise`   Same as above but the running average is
+                                                                        computed across batches in an epoch so it
+                                                                        is reset at the end of the epoch. Default.
+                :class:`~.metrics.metric.RunningEpochWise`              Running average of the ``src`` metric or
+                                                                        ``engine.state.output`` is computed across
+                                                                        epochs. In the former case, ``src`` works
+                                                                        as if it was attached in a
+                                                                        :class:`~ignite.metrics.metric.EpochWise`
+                                                                        manner and its computed value is retrieved
+                                                                        at the end of the epoch. The latter case
+                                                                        doesn't make much sense for this usage as
+                                                                        the ``engine.state.output`` of the last
+                                                                        batch is retrieved then.
+                ======================================================= ===========================================
+
+        ``RunningAverage`` retrieves ``engine.state.output`` at ``usage.ITERATION_COMPLETED`` if the ``src`` is not
+        given and it's computed and updated using ``src`, by manually calling its ``compute`` method, or
+        ``engine.state.output`` at ``usage.COMPLETED``event.
+        Also if ``src`` is given, it is updated at ``usage.ITERATION_COMPLETED``, but its reset event is determined by
+        ``usage`` type. If ``isinstance(usage, BatchWise)`` holds true, ``src`` is reset on ``BatchWise().STARTED``,
+        otherwise on ``EpochWise().STARTED`` if ``isinstance(usage, EpochWise)``.
+        """
+        usage = self._check_usage(usage)
+        has_src = hasattr(self, "src") and isinstance(self.src, Metric)
```

Comment:
Let's set `self.src = None` to indicate that source is Metric instead of checking attribute existence

Reply:
We can't simply use `if self.src` in `attach` since `self.src` gets value in `_output_update` when there's no metric. This makes problem in the case of calling `attach` again.

By the way, why `detach` has not been implemented for `RunningAverage`?

Reply:
> We can't simply use if self.src in attach since self.src gets value in _output_update when there's no metric. This makes problem in the case of calling attach again.

OK, let's just define `self.src = None` in the constructor and than `self.src` will be a metric if input `src` is a metric. This will help us to avoid doing `hasattr(self, "src")`, we can just do:
```
isinstance(self.src, Metric)
```

> By the way, why detach has not been implemented for RunningAverage?

Sorry, I do not remember

Reply:
Also why we don't apply `if not engine.has_event_handler` on `engine.add_event_handler(usage.COMPLETED, self.completed, name)` in `metric.py::attach`?

Reply:
Could you please answer this?

Reply:
> Also why we don't apply `if not engine.has_event_handler` on `engine.add_event_handler(usage.COMPLETED, self.completed, name)` in `metric.py::attach`?

Maybe, it is related to https://github.com/pytorch/ignite/pull/326#issuecomment-440463364

Reply:
But the handlers in a metric don't accept args and kwargs. it's just the handler itself.